File: Java/OPSJLib/src/ops/TopicInfoData.java
/**
*
* Copyright (C) 2013, 2020 Lennart Andersson.
*
* This file is part of OPS (Open Publish Subscribe).
*
* OPS (Open Publish Subscribe) is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.

* OPS (Open Publish Subscribe) is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with OPS (Open Publish Subscribe).  If not, see <http://www.gnu.org/licenses/>.
*/

package ops;

import ops.OPSObject;
import configlib.ArchiverInOut;
import configlib.SerializableFactory;
import configlib.Serializable;
import java.io.IOException;

// NOTE. Must be kept in sync with C++
public class TopicInfoData extends OPSObject
{
    public byte TopicInfoData_version = TopicInfoData_idlVersion;

    public static final byte TopicInfoData_idlVersion = 0;
    public String name;
    public String type;
    public String transport;
    public String address;
    public int port;

    // C++ version: std::vector<string> keys;
    public java.util.Vector<String> keys = new java.util.Vector<String>();

    // C++ version: //std::vector<OPSObject*> filters;


    public TopicInfoData()
    {
        super();
        appendType("TopicInfoData");
    }

    public TopicInfoData(Topic topic)
    {
        super();
        appendType("TopicInfoData");
        name = topic.getName();
        type = topic.getTypeID();
        transport = topic.getTransport();
        address = topic.getDomainAddress();
        port = topic.getPort();
        //keys;
    }

    public void serialize(ArchiverInOut archive) throws IOException
    {
        super.serialize(archive);

        if (idlVersionMask != 0) {
            byte tmp = archive.inout("TopicInfoData_version", TopicInfoData_version);
            if (tmp > TopicInfoData_idlVersion) {
                throw new IOException("TopicInfoData: received version '" + tmp + "' > known version '" + TopicInfoData_idlVersion + "'");
            }
            TopicInfoData_version = tmp;
        } else {
            TopicInfoData_version = 0;
        }

        name = archive.inout("name", name);
        type = archive.inout("type", type);
        transport = archive.inout("transport", transport);
        address = archive.inout("address", address);
        port = archive.inout("port", port);
        keys = (java.util.Vector<String>) archive.inoutStringList("keys", keys);
    }

}

File: CSharp/Ops/TopicInfoData.cs
///////////////////////////////////////////////////////////
//  TopicInfoData.cs
//  Implementation of the Class TopicInfoData
//  Created on:      01-jan-2013 
//  Author:
///////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace Ops
{
    // NOTE. Must be kept in sync with C++
	public class TopicInfoData : OPSObject
	{
        public const byte TopicInfoData_idlVersion = 0;
        private byte _TopicInfoData_version = TopicInfoData_idlVersion;
        public byte TopicInfoData_version
        {
            get { return _TopicInfoData_version; }
            set {
                if (value > TopicInfoData_idlVersion)
                {
                    throw new IdlVersionException(
                        "TopicInfoData: received version '" + value + "' > known version '" + TopicInfoData_idlVersion + "'");
                }
                _TopicInfoData_version = value;
            }
        }

        public string name { get; set; }
        public string type { get; set; }
        public string transport { get; set; }
        public string address { get; set; }
        public int port { get; set; }

        // C++ version: std::vector<string> keys;
        private List<string> _keys = new List<string>();
        [Editor(@"System.Windows.Forms.Design.StringCollectionEditor," +
            "System.Design, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a",
            typeof(System.Drawing.Design.UITypeEditor))]
        public List<string> keys { get { return _keys; } set { _keys = value; } }

		public TopicInfoData()
		{
			AppendType("TopicInfoData");
		}

		public TopicInfoData(Topic topic)
		{
			AppendType("TopicInfoData");
			name = topic.GetName();
			type = topic.GetTypeID();
			transport = topic.GetTransport();
			address = topic.GetDomainAddress();
			port = topic.GetPort();
			//keys;
		}

        public override void Serialize(IArchiverInOut archive)
		{
			base.Serialize(archive);

            if (IdlVersionMask != 0)
            {
                TopicInfoData_version = archive.Inout("TopicInfoData_version", TopicInfoData_version);
            }
            else
            {
                TopicInfoData_version = 0;
            }

            name = archive.Inout("name", name);
            type = archive.Inout("type", type);
            transport = archive.Inout("transport", transport);
            address = archive.Inout("address", address);
            port = archive.Inout("port", port);
            _keys = (List<string>)archive.InoutStringList("keys", _keys);
		}

        public override object Clone()
        {
            TopicInfoData cloneResult = new TopicInfoData();
            FillClone(cloneResult);
            return cloneResult;
        }

        public override void FillClone(OPSObject cloneO)
        {
            base.FillClone(cloneO);
            TopicInfoData cloneResult = (TopicInfoData)cloneO;
            cloneResult.TopicInfoData_version = TopicInfoData_version;
            cloneResult.name = name;
            cloneResult.type = type;
            cloneResult.transport = transport;
            cloneResult.address = address;
            cloneResult.port = port;
            cloneResult.keys = new List<string>(keys);
        }
    };

}
